analysis/topology/sequences.lean

```diff
@@ -0,0 +1,282 @@
+
+universes u v
+variables {X : Type u} {Y : Type v}
+
+def to_filter (x : ℕ → X) : filter X := filter.map x at_top
+
+/- Statements about sequences in general topological spaces. -/
+section topological_space
+variables [topological_space X] [topological_space Y]
+
+/- The notion of convergence of sequences in topological spaces. -/
+def converges_to (x : ℕ → X) (limit : X) : Prop :=
+  ∀ U : set X, limit ∈ U → is_open U → ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U
+
+lemma const_seq_conv (p : X) : converges_to (λ n, p) p :=
+assume U (_: p ∈ U) (_ : is_open U), exists.intro 0 (assume n (_ : n ≥ 0), ‹p ∈ U›)
+
+/- A sequence converges if and only if the associated statement for filter holds. -/
+lemma converges_to_iff_tendsto [topological_space X] {x : ℕ → X} {limit : X} :
+  converges_to x limit ↔ tendsto x at_top (nhds limit) := 
+  iff.intro
+    (assume xtol : converges_to x limit,
+      suffices ∀ U, limit ∈ U → is_open U → x ⁻¹' U ∈ at_top.sets, from tendsto_nhds this,
+        assume U limitInU isOpenU,
+          suffices ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, by simp[this],
+          xtol U limitInU isOpenU)
+    (assume ttol : tendsto x at_top (nhds limit),
+      show ∀ U : set X, limit ∈ U → is_open U → ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from
+        assume U limitInU isOpenU,
+          have {n | (x n) ∈ U} ∈ at_top.sets, from mem_map.mp $ le_def.mp ttol U 
+                                               (mem_nhds_sets isOpenU limitInU),
+          show ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from mem_at_top_sets.mp this)
+
+/- The sequential closure of a subset M ⊆ X of a topological space X is 
+ - the set of all p ∈ X which arise as limit of sequences in M.
+ -/
+def sequential_closure (M : set X) : set X :=
+  {p | ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ M)) ∧ (converges_to x p)}
+
+
+lemma subset_seq_closure (M : set X) : M ⊆ sequential_closure M := 
+assume p (_ :p ∈ M), show p ∈ sequential_closure M, from
+  exists.intro (λ n, p) ⟨assume n, ‹p∈ M›, const_seq_conv p⟩
+
+def is_seq_closed (A : set X) : Prop := A = sequential_closure A
+
+/- A convenience lemma for showing that a set is sequentially closed. -/
+lemma is_seq_closed_of_def {A : set X} (h : ∀ (x : ℕ → X), (∀ n : ℕ, ((x n) ∈ A)) → ∀ p : X,
+  converges_to x p → p ∈ A) : is_seq_closed A :=
+show A = sequential_closure A, from set.ext (assume p, iff.intro
+(assume : p ∈ A, subset_seq_closure A ‹p ∈ A›)
+(assume :p ∈ sequential_closure A, 
+  have ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ A)) ∧ (converges_to x p), by assumption,
+  let ⟨x, ⟨_, _⟩⟩ := this in
+  show p ∈ A, from h x ‹∀ n : ℕ, ((x n) ∈ A)› p ‹converges_to x p›))
+
+/- The sequential closure of a set is containt in the closure of that set. The converse is not
+ - true.
+ -/ 
+lemma sequential_closure_subset_closure (M : set X) : sequential_closure M ⊆ closure M :=
+suffices ∀ p, p ∈ sequential_closure M → p ∈ closure M, by assumption,
+assume p,
+assume : ∃ x : ℕ → X, (∀ n :ℕ, ((x n) ∈ M)) ∧ (converges_to x p),
+let ⟨x, ⟨_, _⟩⟩ := this in
```

Comment:
You can combine these two lines: `assume ⟨x, ⟨_, _⟩⟩ : ...`

Reply:
How? If I write 
`assume ⟨x, ⟨_, _⟩⟩ : ∃ x : ℕ → X, (∀ n :ℕ, ((x n) ∈ M)) ∧ (converges_to x p),`
lean complains "command expected, invalid binder". This somehow makes sense to me, to actually pick an x, I need choice.

Reply:
Oh, sorry, I didn't check the syntax before I wrote that. Apparently you can only do this anonymously. `assume ⟨x, ⟨_, _⟩⟩, ...` without the type should work, but I guess that doesn't fit the verbose style you're using.

Choice doesn't come into play here in either case. You're using a prop to prove a prop; the `let` is just pattern matching on the proof of the exists and getting an `x` from that. What I wrote is just alternate syntax for what you wrote (except for the lack of an explicit type, which I just assumed was supported).

Reply:
Yep, then let's leave it this way.